### PR TITLE
Add Zod for input validation on LoginScreen and ManualVoucherScreen

### PR DESCRIPTION
**Pull request by ronniebeggs**
[//]: # "These comments are meant for your reference. They are invisible and don't need to be deleted!"

### Description
[//]: # "Required - Describe what's new in this PR in a few lines. A description and bullet points for specifics will suffice."
- Added basic input validation to input fields on the login and manual voucher input screens
- Uses ZOD library to parse inputs and guarantee proper types/ minimum input lengths
- Validation functions triggered when the user unfocuses from an input field
- Added a prop to the input field component that triggers a different keypad type depending on the input
- Added regex to onChange callbacks for input filtering on android

### Screenshots
[//]: # "Required for frontend changes, otherwise optional but strongly recommended. Add screenshots of expected behavior - GIFs if you're feeling fancy!"

Image Order of Keyboard Types: Number --> Decimal --> Email
  

https://user-images.githubusercontent.com/66931067/223201963-30c6730f-6bfe-4ef3-87bc-3d6b888e3498.MP4

In the demo, an invalid email expression triggers an error.

## How to review
[//]: # "Required - Describe the order in which to review files and what to expect when testing locally. Is there anything specifically you want feedback on? Should this be reviewed commit by commit, or all at once? What are some user flows to test? What are some edge cases to look out for?"
- Try various inputs in the login input fields, and watch as errors populate the console
- Login and navigate to the manual voucher input screen
- Similarly try various invalid inputs and observe errors
- Notice that inputs fields use different keyboard types

## Next steps
[//]: # "Optional - What's NOT in this PR, doesn't work yet, and/or still needs to be done. Note any temporary fixes in this PR that should be cleaned up later."
- Catch and display errors for the user 
- Create a more comprehensive input validation procedure for currency inputs

### Online sources
[//]: # "Optional - copy links to any tutorials or documentation that was useful to you when working on this PR"
[ZOD Documentation](https://zod.dev/?id=numbers)
[React Native TextInput Keyboard Types](https://lefkowitz.me/visual-guide-to-react-native-textinput-keyboardtype-options/)

[//]: # "This tags the project leader as a default. Feel free to change, or add on anyone who you should be in on the conversation."
🥦 CC: @sauhardjain

### CODE REVIEW

**Comment by selene-huang** [comment]
Excellent work Ronnie! Some issues I noticed while testing:

- Transaction ID has the option to put both decimal points and - signs on my keyboard, these should probably be disabled!
- Voucher amount had the - sign disabled but not the decimal point - I recall seeing a discussion about displaying things as a dollar amount, but forcing users to input as cents?
- Seems like you're validating transaction ID & voucher amount both as strings. Given the input types you've provided, this is actually the same as having no validation! I think it would make sense for them to be evaluated as integers instead.

I would give feedback on the login screen too, but I was actually completely unable to open the login screen on my phone for testing. Since I can't open it on main either, I don't think this has anything to do with your PR though.

**Review by sauhardjain — CHANGES_REQUESTED** [review]
Great work so far! Some other bugs to tackle in addition to the ones Selene pointed out:
- [x] Add a max limit of 10 dollars to the voucher value input box.
- [x] Find a way to add a Done button to the numeric keyboards because currently there's no way to close these keyboards.

IMG_BD1EDF4669A9-1

- [x] On the login screen, Zod simultaneously throws an error for both the email and password even if the user has only inputted one out of the two incorrectly.

IMG_EDE1F83F60E3-1

File: src/screens/scanning/ManualVoucherScreen.tsx [hunk 1]
```diff
@@ -18,6 +20,8 @@ import {
 import InputField from '../../components/InputField/InputField';
 import StandardLogo from '../../components/common/StandardLogo';
 
+Logs.enableExpoCliLogging();
```
Comment (wangannie) [inline src/screens/scanning/ManualVoucherScreen.tsx]: Remove?

File: src/screens/scanning/ManualVoucherScreen.tsx [hunk 2]
```diff
@@ -30,6 +34,45 @@ function ManualVoucherScreen() {
   //   { label: 'Banana', value: 'banana' },
   // ]);
 
+  const SNonChange = (text: string) => {
```
Comment (wangannie) [inline src/screens/scanning/ManualVoucherScreen.tsx]: Rename to `onChangeSerialNumber`

File: src/screens/scanning/ManualVoucherScreen.tsx [hunk 3]
```diff
@@ -30,6 +34,45 @@ function ManualVoucherScreen() {
...
+    const value = text.replace(/\D/g, '');
+    setID(value);
+  };
+
+  const VAonChange = (text: string) => {
```
Comment (wangannie) [inline src/screens/scanning/ManualVoucherScreen.tsx]: Rename to `onChangeVoucherAmount`

File: src/screens/scanning/ManualVoucherScreen.tsx [hunk 4]
```diff
@@ -30,6 +34,51 @@ function ManualVoucherScreen() {
...
+    try {
+      const SNSchema = z.coerce.number().gt(0);
+      SNSchema.parse(transactionID);
+    } catch (error) {
+      // eslint-disable-next-line no-console
+      console.log(error);
+    }
+  };
+
+  const validateCurrencyFormat = (input: string) => {
```
Comment (wangannie) [inline src/screens/scanning/ManualVoucherScreen.tsx]: I would move these validation functions to a separate util file like `validationUtils.ts` or `voucherUtils.ts` in the `src/utils` folder.

File: src/components/InputField/InputField.tsx [hunk 5]
```diff
@@ -30,6 +37,8 @@ export default function InputField({
       secureTextEntry={secureTextEntry}
       autoCorrect={false}
       autoCapitalize="none"
+      keyboardType={inputMode}
```
Comment (wangannie) [inline src/components/InputField/InputField.tsx]: `inputMode` is also a prop on `TextInput` right? you could probably use that directly

File: src/screens/auth/LoginScreen.tsx [hunk 6]
```diff
@@ -51,6 +52,20 @@ export default function LoginScreen({
...
 
+  const validateEmailInput = () => {
+    if (email !== '') {
+      const emailSchema = z.string().email();
+      emailSchema.parse(email);
+    }
+  };
+
+  const validatePasswordInput = () => {
+    if (password !== '') {
```
Comment (wangannie) [inline src/screens/auth/LoginScreen.tsx]: We're removing these checks right?

File: src/screens/auth/LoginScreen.tsx [hunk 7]
```diff
@@ -51,6 +52,20 @@ export default function LoginScreen({
     setPassword(value);
   };
 
+  const validateEmailInput = () => {
```
Comment (wangannie) [inline src/screens/auth/LoginScreen.tsx]: Similarly I would add these to a file like `validationUtils`

**Review by wangannie — CHANGES_REQUESTED** [review]
Looks really good! Mostly noted thoughts on code cleanup. They should be quick fixes : )

File: src/components/InputField/InputField.tsx [hunk 8]
```diff
@@ -30,6 +37,8 @@ export default function InputField({
       secureTextEntry={secureTextEntry}
       autoCorrect={false}
       autoCapitalize="none"
+      keyboardType={keyboard}
```
Comment (wangannie) [inline src/components/InputField/InputField.tsx]: nit: for consistency, in most cases we'd wanna match the name of the props we're overriding so `keyboardType` instead of `keyboard`

File: src/utils/validationUtils.ts [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    const SNSchema = z.coerce.number().gt(0);
+    SNSchema.parse(input);
+  } catch (error) {
+    // eslint-disable-next-line no-console
+    console.log(error);
+  }
+};
+
+const validateCurrencyFormat = (input: string) => {
+  // confirms that any input has either 0 or 2 decimal digits
```
Comment (wangannie) [inline src/utils/validationUtils.ts]: I like that you included a comment! Super helpful. I would also just move this to a docstring so it shows up nicely in VSCode

```
/**
 * .. your comment...
 */
```

**Review by wangannie — APPROVED** [review]
LOOKS GREAT! A few more nits, but good to merge after!

**Review by sauhardjain — APPROVED** [review]
